### ISSUE TITLE
Multiple issues with fairly complicated schematron

### ISSUE THREAD

**Issue by ahenket** [phax/ph-schematron#42]
Hi. This relates to at least issue#41. I've got a package to reproduce with. Note that running the very same thing inside Oxygen XML editor which also carries Saxon works and gives 23 correct errors.

Help appreciated

Issues:

1. document() function not supported. The sch leverages plenty of document() calls to files in an include directory right next to the main schematron file, but unless I convert those into doc() calls I get a compile error on every single one of those.

2. Relative paths do not work. Unless I place the sch + include + instance inside my Java project folder structure it cannot resolve the relative paths that the document() function calls

3. Bogus errors. It looses context on asserts called from abstract rules leading to bogus errors. Code I used is this:

```
	File file = new File("totalindberetning-eksempel.xml");
	DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
	DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
	Document document = documentBuilder.parse(file);

	String sch = "lpr-lpr3.sch";

	SchematronResourcePure aResPure = SchematronResourcePure.fromFile(sch);
	SchematronOutputType validationResult;

	validationResult = aResPure.applySchematronValidationToSVRL(document);
	System.out.println("ActivePatternAndFiredRuleAndFailedAssertCount "
			+ validationResult.getActivePatternAndFiredRuleAndFailedAssertCount());
	for (int i = 0; i < validationResult.getActivePatternAndFiredRuleAndFailedAssertCount(); i++) {
		Object obj = validationResult.getActivePatternAndFiredRuleAndFailedAssertAtIndex(i);

		if (obj instanceof FailedAssert) {
			FailedAssert fa = (FailedAssert) obj;

			System.out.print(i + ". ");
			System.out.println(fa.getText());
			System.out.print("    ");
			System.out.println(fa.getLocation());
			System.out.print("    ");
			System.out.println(fa.getTest());
		}
	}
```

The last of the 2260 errors I get is this one:

```
2260. dtr1-18-CD: [RIM-002] code (@code= @codeSystem=) SHALL be distinct
    /ClinicalDocument/component/structuredBody/component[2]/section/component/procedure/code
    not(preceding-sibling::*[local-name()=local-name()] | following-sibling::*[local-name()=local-name()][@code=@code][@codeSystem=@codeSystem])
```

You'll see a lot more similar errors and each of those is simply incorrect. By the looks of it, it looses context, for example because @code= @codeSystem= would have had values if the assertion text was properly evaluated.

`    dtr1-18-CD: [RIM-002] <value-of select="$name"/> (@code=<value-of select="$code"/> @codeSystem=<value-of select="$codeSystem"/>) SHALL be distinct`

[reproduction-package.zip](https://github.com/phax/ph-schematron/files/993998/reproduction-package.zip)

**Comment by phax** [comment]
Hi! Will look at it next week.
Short question: what happens when you change from `SchematronResourcePure` to `SchematronResourceSCH`?

Thanks,
  Philip

**Comment by ahenket** [comment]
Wow that solved a number of things:

1. Solved: document() is supported as expected

2. Not solved: relative paths are still not resolved correctly. It assumes relative to class file instead of relative to the calling schematron file

3. Solved: now returns the 23 errors I expect to get

So that brings it down to bigger problems in `SchematronResourcePure` and a relatively small problem in `SchematronResourceSCH`

**Comment by phax** [comment]
That is exactly the difference in the implementations. The "pure" implementation supports only XPath whereas the `SCH` implementation uses the XSLT way to convert it to XSLT and therefore all the XSLT methods are available.

As a performance improvement I suggest you convert the SCH files once to XSLT and just apply the "SCH-XSLT" onto the XML to be validated - it will create an SVRL.

Ad 2) In versions up to 4.2.2 you can either set an `EntityResolver` for the "pure" implementation or an `URIResolver` for the "SCH" version - the pure version uses it for Schematron and for XML, whereas the "SCH" version uses it only for the Schematron. In Version 4.3.0 I added the possibility to also set an `EntityResolver` for "SCH" mode so that it can be used for the XML versions to be validated.

So do you think it would make sense to set a default "simple" URIResolver and EntityResolver ???

**Comment by ahenket** [comment]
> The "pure" implementation supports only XPath 

Is there a version missing like 1.0 here? I do not understand this part of the sentence.

> So do you think it would make sense to set a default "simple" URIResolver and EntityResolver ???

I'm responsible for producing the schematrons and we also provide a way to use them. We actually also produce SVRL normally. One of our users @jkiddo, who opened issue #41, asked me to help analyse, but I'm not a skilled Java programmer so I'm not sure how to respond here. I hope Jens will step here to assist you in deciding the best way forward.

**Comment by phax** [comment]
@ahenket XPath is a "subset" of XSLT (dear XML gurus - don't kill me for that). Basic meaning: it is limited but therefore quicker. If you have Schematrons that are optimized for speed and don't use XSLT features the "pure" implementation is the tool of choice.

Got your point - I think I will provide a reasonable default - it's just up to @jkiddo to test when I'm ready ;-)

**Comment by ahenket** [comment]
What would you consider an XSLT feature? One of the things I suspect are not working are the rule context paths causing assertion tests to be execute in the wrong context. Also <let/> statements do not seem to be used. Both should be within XPath are they not?

**Comment by phax** [comment]
E.g. `document()` is an XSLT feature, `doc()` is an XPath feature. XSLT uses XPath for finding elements ("path to XML node" - x-path) but offers more like variables, functions etc. XSLT is even a turing complete programming language wheres XPath is not.

To use a function in XPath that is not standard, you need to define it. Therefore you have the possibility to add a `FunctionResolver` but I think this is again too deep in the Java code.
So theoretically it is possible to use `document()` with the "pure" implementation, but than you need to provide your own implementation :)

**Comment by ahenket** [comment]
Gotcha. I'll let Jens jump in here.

**Comment by phax** [comment]
So - the latest version of 4.3.0-SNAPSHOT contains an EntityResolver and a URIResolver by default. The try to resolve all stuff relative to the Schematron file. hth

**Comment by jkiddo** [comment]
Wow - I'm really sure that I had tried the SCH version with the same results, but I can now see, that I must have done it wrong. Depending on the evaluation speed, the SCH-version could be a solution to me - but I would require the relative file location to be fixed.

**Comment by jkiddo** [comment]
Anyways - @ahenket and @phax - thank you very much for your help! You guys seems to have saved the day

**Comment by phax** [comment]
@jkiddo Please try the latest 4.3.0-SNAPSHOT - it has some EntityResolver/URIResolver built-in.

**Comment by phax** [comment]
4.3.0 is out. Please check https://github.com/phax/ph-schematron/releases/tag/ph-schematron-parent-pom-4.3.0
If the prolem with the EntityResolver remains please open a new issue. thx
